chore: add script to pull commits to cherrypick

andrewseguin opened the pull request:
Example output:

```
^_^ node scripts/caretaking/cherrypick-patch.js 2019-07-20
Loading 11 pull requests
Loading PR #16604
Loading PR #16593
Loading PR #16587
Loading PR #16574
Loading PR #16571
Loading PR #16561
Loading PR #16556
Loading PR #16545
Loading PR #16466
Loading PR #16091
Loading PR #16057

16561	2019-07-22T18:41:02Z	701302dc482d7e4b77990b24e3b5ab330bbf1aa5	https://github.com/angular/components/pull/16561	style(mdc chips) Fix clashing input and remove classes
16604	2019-07-24T23:52:23Z	b5b9ff726e984b1ba31febea30b5afa3153ebbeb	https://github.com/angular/components/pull/16604	fix(popover-edit): blending in with background in high contrast mode
16593	2019-07-24T23:54:28Z	583af19e9cc6fb37e1e4d32fe2abcf9700d12fff	https://github.com/angular/components/pull/16593	feat(material-experimental): add test harness for radio-button
16571	2019-07-24T23:54:59Z	b0991de9284f5f24f708dcd92dca034623cf7595	https://github.com/angular/components/pull/16571	docs(expansion): fix docs for togglePosition input
16057	2019-07-24T23:58:40Z	13becdab50b6b9cf979ef9b3745f4fa29ff750c7	https://github.com/angular/components/pull/16057	fix(popover-edit): incorrectly caching values for more than one control
16091	2019-07-24T23:59:18Z	174bf4076a5c0a97d922741230c5e8143c0dc84f	https://github.com/angular/components/pull/16091	fix(popover-edit): focus trap not being destroyed and potential leaks with takeUntil
16587	2019-07-25T00:03:46Z	bd24369d3883cee834662800b0a5b043ffdc1c6c	https://github.com/angular/components/pull/16587	fix(button-toggle): error when checked value is set via static attribute in Ivy
16574	2019-07-25T00:04:38Z	67532dbbb8a14dfdca2134b506b6269d76700516	https://github.com/angular/components/pull/16574	fix(mdc chips) Fix matChipRemove Expression has changed error
16466	2019-07-25T18:39:31Z	615070acbdb2962eca3bebadb19720c614848981	https://github.com/angular/components/pull/16466	fix(popover-edit): unable to close by pressing enter on some screen readers
16545	2019-07-25T20:14:19Z	2a04ccfd46c154cc5a23993cda8dff3474f2248c	https://github.com/angular/components/pull/16545	feat(material-experimental): add test harnesses for slide-toggle
16556	2019-07-26T15:56:47Z	88631b936d371b6157cb686a88b2fef35c30dc0b	https://github.com/angular/components/pull/16556	feat(material-experimental): add test harness for button
---------------
--Cherry-pick--
---------------

git cherry-pick 701302dc482d7e4b77990b24e3b5ab330bbf1aa5;
git cherry-pick b5b9ff726e984b1ba31febea30b5afa3153ebbeb;
git cherry-pick 583af19e9cc6fb37e1e4d32fe2abcf9700d12fff;
git cherry-pick b0991de9284f5f24f708dcd92dca034623cf7595;
git cherry-pick 13becdab50b6b9cf979ef9b3745f4fa29ff750c7;

git cherry-pick 174bf4076a5c0a97d922741230c5e8143c0dc84f;
git cherry-pick bd24369d3883cee834662800b0a5b043ffdc1c6c;
git cherry-pick 67532dbbb8a14dfdca2134b506b6269d76700516;
git cherry-pick 615070acbdb2962eca3bebadb19720c614848981;
git cherry-pick 2a04ccfd46c154cc5a23993cda8dff3474f2248c;

git cherry-pick 88631b936d371b6157cb686a88b2fef35c30dc0b;
```

mmalerba commented on scripts/caretaking/cherrypick-patch.js:
```diff
@@ -0,0 +1,88 @@
+ */
+
+/** Required argument: Date of the oldest PR to load, should be in the form of YYYY-MM-DD. */
+const date = process.argv[2];
```
Could you just save the last cp'd sha in some file we check in to source control as the default? That way the caretaker could run this as part of doing the release and just check in the updated sha along with the changes to the CHANGELOG.md and package.json

mmalerba commented on package.json:
```diff
@@ -71,7 +71,7 @@
     "@bazel/typescript": "0.32.2",
     "@firebase/app-types": "^0.3.2",
-    "@octokit/rest": "^15.9.4",
+    "@octokit/rest": "^16.28.7",
```
you should add a script to package.json so we can do `yarn cherrypick` or something

mmalerba commented on scripts/caretaking/cherrypick-patch.js:
```diff
@@ -0,0 +1,88 @@
+}
+
+/** Optional argument: Github personal access token for increased API rate limit. **/
+const authToken = process.argv[3];
```
is there a good way to read system environment vars in node? that way we could each export something in our bashrc, i.e. GIT_TOKEN

jelbourn commented on scripts/caretaking/cherrypick-patch.js:
```diff
@@ -0,0 +1,88 @@
+const authToken = process.argv[3];
+
+const Octokit = require('@octokit/rest');
+const octokit = new Octokit({auth: authToken});
```
I would call this `github`

jelbourn commented on scripts/caretaking/cherrypick-patch.js:
```diff
@@ -0,0 +1,88 @@
+  const numbers = [];
+  for (const pr of response.data.items) {
+    numbers.push(pr.number);
+  }
```
```ts
const numbers = response.data.items.map(pr => pr.number);
```

jelbourn commented on scripts/caretaking/cherrypick-patch.js:
```diff
@@ -0,0 +1,88 @@
+  for (let i = 0; i < numbers.length; i++) {
+    console.log(`Loading PR #${numbers[i]}`);
+    pullRequests.push(await loadPullRequestData(owner, repo, numbers[i]));
+  }
```
```ts
const pullRequests = numbers.map(n => await loadPullRequestData(owner, repo, n));
```
You could also just omit `numbers`
```ts
const pullRequests = 
    response.data.items.map(pr => await loadPullRequestData(owner, repo, pr.number));
```

Alternatively, it would probably be faster to not use `await` in the loop and just `push` in a `then` 

(I would just omit the specific "loading..." message)

devversion commented on scripts/caretaking/cherrypick-patch.js:
```diff
@@ -0,0 +1,88 @@
+  const query =
+      `repo:${owner}/${repo} base:master is:pr is:merged merged:>${date} ${excludeLabelsQueryStr}`;
+
+  const response = await octokit.search.issuesAndPullRequests({per_page: 100, q: query});
```
Even though we rarely might hit that, it looks like this will only work for `100` PRs merged. 

If we ever would hit that it could be difficult to see that some PRs were missed. should we use pagination?

andrewseguin commented:
- Moved the script to `tools` and changed it to be Typescript
- Added command from package.json: `yarn cherry-pick-patch`
- Attempts to automatically default to the latest patch branch, prompts user to confirm
- Uses the patch branch's commit messages to find the latest SHA to cherry-pick
- Confirms with user if the commit is correct

TODO's left for another PR:
 - Handle pagination for cases were there are over 100 commits to cherry-pick (thanks Paul)
 - Use a local environment variable for authentication token (thanks Miles)

jelbourn commented:
Lint error
